scheds/rust/scx_p2dq/src/bpf/main.bpf.c

```diff
@@ -1065,47 +1023,122 @@ static __always_inline int dispatch_cpu(u64 dsq_id, s32 cpu, struct llc_ctx *llc
 	return dispatched;
 }
 
+static __always_inline bool consume_llc_compat(struct llc_ctx *cur_llcx, struct llc_ctx *llcx)
+{
+	u64 dsq_id;
+	int i;
 
-static __always_inline int dispatch_pick_two(s32 cpu, struct llc_ctx *cur_llcx, struct cpu_ctx *cpuc)
+	if (dispatch_lb_interactive &&
+	    scx_bpf_dsq_move_to_local(llcx->dsqs[0])) {
+			stat_inc(P2DQ_STAT_DISPATCH_PICK2);
+			return true;
+	}
+
+	if (llcx->load > cur_llcx->load) {
+		bpf_for(i, 1 , nr_dsqs_per_llc) {
+			dsq_id = llcx->dsqs[nr_dsqs_per_llc - i];
+			if (scx_bpf_dsq_move_to_local(dsq_id)) {
+				stat_inc(P2DQ_STAT_DISPATCH_PICK2);
+				return true;
+			}
+		}
+	}
+
+	return false;
+}
+
+static __always_inline bool consume_llc(struct llc_ctx *cur_llcx, struct llc_ctx *llcx, s32 cpu)
 {
-	struct llc_ctx *llcx, *left, *right;
 	u64 dsq_id;
 	int i;
 
+	if (!cur_llcx || !llcx || !bpf_ksym_exists(scx_bpf_dsq_move))
+		return false;
+
+	if (dispatch_lb_interactive &&
+	    scx_bpf_dsq_move_to_local(llcx->dsqs[0])) {
+		stat_inc(P2DQ_STAT_DISPATCH_PICK2);
+		return true;
+	}
+
+	bpf_for(i, 1, nr_dsqs_per_llc) {
+		dsq_id = llcx->dsqs[i];
```

Comment:
You shouldn't iterate in reverse order here (`nr_dsqs_per_llc - i` instead of `i`), like you do in `consume_llc_compat()`? Maybe I'm missing something.

Reply:
Good catch, I was trying something out with reverse iteration and accidentally left this.